runtime/src/comm/gasnet/comm-gasnet-ex.c

```diff
@@ -1084,23 +1066,15 @@ void chpl_comm_broadcast_private(int id, size_t size) {
 }
 
 void chpl_comm_impl_barrier(const char *msg) {
-  int id = (int) msg[0];
-  int retval;
-
-  //
-  // We don't want to just do a gasnet_barrier_wait() here, because
-  // GASNet will put us to work polling, and we already have a polling
-  // task that the tasking layer has presumably placed to best effect.
-  // We don't want to compete with that.  Also, the implementation is
-  // required to do chpl_task_yield() while waiting for the barrier to
-  // satisfy; see chpl_comm.h.  This prevents us from monopolizing the
-  // processor while waiting.
-  //
-  gasnet_barrier_notify(id, 0);
-  while ((retval = gasnet_barrier_try(id, 0)) == GASNET_ERR_NOT_READY) {
+  gex_Event_t bar = gex_Coll_BarrierNB(myteam, GEX_NO_FLAGS);
+  while (gex_Event_Test(bar) != GASNET_OK) {
+    // Need to poll if polling is required and progress thread isn't running
+    // (polling thread is started late in init and stopped early in deinit)
+    if (pollingRequired && !pollingRunning) {
+      gasnet_AMPoll();
+    }
     chpl_task_yield();
   }
```

Comment:
I suggest polling unconditionally in this loop:

```suggestion
  gex_Event_t bar = gex_Coll_BarrierNB(myteam, GEX_NO_FLAGS);
  while (gex_Event_Test(bar) != GASNET_OK) {
    // ibv-conduit polling thread alone does not reliably advance barriers,
    // so poll here unconditionally
    gasnet_AMPoll();
    chpl_task_yield();
  }
```

This will probably fix your hang, although it might increase lock contention and hence injection overhead for concurrent threads.

Reply:
Yeah, an unconditional poll resolves the hang. I was hoping to avoid that with the goal of reducing contention, but seems like it's needed. I expect to use our `am_poll_try` wrapper though.

Do you have a sense for what about the RDMA-based barrier requires AM progress calls?

Reply:
> Do you have a sense for what about the RDMA-based barrier requires AM progress calls?

It's not that it requires AM progress calls, because it doesn't use AM. `gasnet_AMPoll()` has a slightly misleading name because it also polls the GASNet barrier subsystem (along with a few other GASNet subsystems), and when RDMADISSEM barrier is active that includes checking whether any incoming RMA Puts have arrived to advance the barrier state.

Reply:
Understood, thanks. And no current or expected interface to progress specific subsystems (I believe know, but want to verify).

Already captured as "explore which functions poll internally" in https://github.com/chapel-lang/chapel/issues/14443, but I think it'll be important to consider where progress calls are needed as some things are converted to use events (and as we look to make Blocking RMA calls into non-blocking ones)

Reply:
> And no current or expected interface to progress specific subsystems (I believe no, but want to verify).

Correct.

We expect there will eventually be a mechanism to progress a specific set of endpoints, but we're not there yet (and currently Chapel is only using one endpoint per process).